### PR TITLE
Filter preload paths only for translatable entities.

### PR DESCRIPTION
**Pull request by Hug0-Drelon**
## What?
- Add language parameter in preload paths only for translatable entities.

## Why?
- Since it's already done in Pro version and in order to avoid side effects.

## How?
- Create `PLL_Filter_REST_Routes` which allow to add query parameters on preload paths only for translatable entities.
- It also print inline script to retrieve filterable routes on client side.

### CODE REVIEW

File: admin/admin-block-editor.php [hunk 1]
```diff
@@ -15,56 +15,77 @@ class PLL_Admin_Block_Editor {
...
-	 *
+	 * @var PLL_CRUD_Posts
+	 */
+	protected $posts;
+
+	/**
 	 * @var PLL_Language|null
 	 */
-	protected $pref_lang;
+	protected $curlang;
```
Comment (manooweb) [inline admin/admin-block-editor.php:25]: Is it useful?

File: phpstan-baseline.neon [hunk 2]
```diff
@@ -51,7 +51,7 @@ parameters:
 			path: admin/admin-base.php
 
 		-
-			message: "#^Cannot access property \\$slug on PLL_Language\\|null\\.$#"
+			message: "#^Property PLL_Admin_Block_Editor\\:\\:\\$posts \\(PLL_CRUD_Posts\\) does not accept PLL_CRUD_Posts\\|null.$#"
```
Comment (manooweb) [inline phpstan-baseline.neon:54]: Shouldn't we document `$posts` property as it could be `null`?

**Review by manooweb — CHANGES_REQUESTED** [review]

File: admin/admin-block-editor.php (hunk 1 above)
Comment (Hug0-Drelon) [reply admin/admin-block-editor.php]: Not at all, remove in e53cc4a9.

File: phpstan-baseline.neon (hunk 2 above)
Comment (Hug0-Drelon) [reply phpstan-baseline.neon]: I thought it would be overkill, but why not.
See f56a40ea

File: tests/phpunit/includes/testcase-preload-paths.php [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+	 */
+	protected function call_sut( $paths, $context ) {
+		return apply_filters( 'block_editor_rest_api_preload_paths', $paths, $context );
+	}
+
+	/**
+	 * @dataProvider preload_paths_provider
+	 *
+	 * @param string|array $path            The preload path under test. Could be an array if provided along a HTTP method.
+	 * @param bool         $is_filtered     Whether the pass should be filtered or not.
```
Comment (manooweb) [inline tests/phpunit/includes/testcase-preload-paths.php:64]: Path, isn't it?
```suggestion
	 * @param bool         $is_filtered     Whether the path should be filtered or not.
```

File: tests/phpunit/includes/testcase-preload-paths.php [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+	 */
+	abstract public function test_preload_paths_with_post_editor_context( $path, $is_filtered, $post, $language, $is_translatable );
+
+	abstract public function test_preload_path_for_translatable_media();
+
+	/**
+	 * @dataProvider preload_paths_provider
+	 *
+	 * @param string|array $path            The preload path under test. Could be an array if provided along a HTTP method.
+	 * @param bool         $is_filtered     Whether the pass should be filtered or not.
```
Comment (manooweb) [inline tests/phpunit/includes/testcase-preload-paths.php:77]: copy/paste drama 😉

File: tests/phpunit/includes/testcase-preload-paths.php [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+	 * @param string       $language        The post's language slug, empty if none.
+	 * @param bool         $is_translatable Whether or not the post type is translatable.
+	 */
+	abstract public function test_preaload_paths_in_site_editor_context( $path, $is_filtered, $post, $language, $is_translatable );
+
+	/**
+	 * @dataProvider preload_paths_provider
+	 *
+	 * @param string|array $path            The preload path under test. Could be an array if provided along a HTTP method.
+	 * @param bool         $is_filtered     Whether the pass should be filtered or not.
```
Comment (manooweb) [inline tests/phpunit/includes/testcase-preload-paths.php:88]: copy/paste drama 😉

File: tests/phpunit/includes/testcase-preload-paths.php [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+	}
+
+	/**
+	 * Provides preload paths with different dataset combination such as:
+	 *     - Standard, translatable or unstranslatable custom post type.
+	 *     - Default, secondary or undefined language for a translatable post.
+	 *
+	 * @return array $data {
+	 *    @type string|array $path            The preload path under test. Could be an array if provided along a HTTP method.
+	 *    @type bool         $is_filtered     Whether the pass should be filtered or not.
```
Comment (manooweb) [inline tests/phpunit/includes/testcase-preload-paths.php:142]: copy/paste drama 😉

File: tests/phpunit/tests/test-preload-paths.php [hunk 7]
```diff
@@ -0,0 +1,99 @@
+<?php
+
+class Preload_Paths_Test extends PLL_Preload_Paths_TestCase {
+	/**
+	 * @dataProvider preload_paths_provider
+	 *
+	 * @param string|array $path            The preload path under test. Could be an array if provided along a HTTP method.
+	 * @param bool         $is_filtered     Whether the pass should be filtered or not.
```
Comment (manooweb) [inline tests/phpunit/tests/test-preload-paths.php:8]: copy/paste drama 😉

File: tests/phpunit/tests/test-preload-paths.php [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+		} else {
+			$this->assertStringNotContainsString( "lang={$language}", $filtered_path, "{$expected_path} should not have the language parameter added." );
+		}
+	}
+
+	/**
+	 * @dataProvider preload_paths_provider
+	 *
+	 * @param string|array $path            The preload path under test. Could be an array if provided along a HTTP method.
+	 * @param bool         $is_filtered     Whether the pass should be filtered or not.
```
Comment (manooweb) [inline tests/phpunit/tests/test-preload-paths.php:52]: copy/paste drama 😉

File: tests/phpunit/tests/test-preload-paths.php [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+	 */
+	public function test_preaload_paths_in_site_editor_context( $path, $is_filtered, $post, $language, $is_translatable ) {
+		$this->assert_unfiltered_path_for_context( $path, 'core/edit-site' );
+	}
+
+	/**
+	 * @dataProvider preload_paths_provider
+	 *
+	 * @param string|array $path            The preload path under test. Could be an array if provided along a HTTP method.
+	 * @param bool         $is_filtered     Whether the pass should be filtered or not.
```
Comment (manooweb) [inline tests/phpunit/tests/test-preload-paths.php:65]: copy/paste drama 😉

File: tests/phpunit/includes/testcase-preload-paths.php [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+	/**
+	 * @dataProvider preload_paths_provider
+	 *
+	 * @param string|array $path            The preload path under test. Could be an array if provided along a HTTP method.
+	 * @param bool         $is_filtered     Whether the pass should be filtered or not.
+	 * @param WP_Post      $post            The post provided for the context.
+	 * @param string       $language        The post's language slug, empty if none.
+	 * @param bool         $is_translatable Whether or not the post type is translatable.
+	 */
+	abstract public function test_preaload_paths_in_widget_editor_context( $path, $is_filtered, $post, $language, $is_translatable );
```
Comment (manooweb) [inline tests/phpunit/includes/testcase-preload-paths.php:93]: Typo

```suggestion
	abstract public function test_preload_paths_in_widget_editor_context( $path, $is_filtered, $post, $language, $is_translatable );
```

File: tests/phpunit/includes/testcase-preload-paths.php [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+	/**
+	 * @dataProvider preload_paths_provider
+	 *
+	 * @param string|array $path            The preload path under test. Could be an array if provided along a HTTP method.
+	 * @param bool         $is_filtered     Whether the pass should be filtered or not.
+	 * @param WP_Post      $post            The post provided for the context.
+	 * @param string       $language        The post's language slug, empty if none.
+	 * @param bool         $is_translatable Whether or not the post type is translatable.
+	 */
+	abstract public function test_preaload_paths_in_site_editor_context( $path, $is_filtered, $post, $language, $is_translatable );
```
Comment (manooweb) [inline tests/phpunit/includes/testcase-preload-paths.php:82]: Typo

```suggestion
	abstract public function test_preload_paths_in_site_editor_context( $path, $is_filtered, $post, $language, $is_translatable );
```

File: tests/phpunit/tests/test-preload-paths.php [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+	 * @param WP_Post      $post            The post provided for the context.
+	 * @param string       $language        The post's language slug, empty if none.
+	 * @param bool         $is_translatable Whether or not the post type is translatable.
+	 */
+	public function test_preaload_paths_in_site_editor_context( $path, $is_filtered, $post, $language, $is_translatable ) {
```
Comment (manooweb) [inline tests/phpunit/tests/test-preload-paths.php:57]: Typo

```suggestion
	public function test_preload_paths_in_site_editor_context( $path, $is_filtered, $post, $language, $is_translatable ) {
```

File: tests/phpunit/tests/test-preload-paths.php [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+	/**
+	 * @dataProvider preload_paths_provider
+	 *
+	 * @param string|array $path            The preload path under test. Could be an array if provided along a HTTP method.
+	 * @param bool         $is_filtered     Whether the pass should be filtered or not.
+	 * @param WP_Post      $post            The post provided for the context.
+	 * @param string       $language        The post's language slug, empty if none.
+	 * @param bool         $is_translatable Whether or not the post type is translatable.
+	 */
+	public function test_preaload_paths_in_widget_editor_context( $path, $is_filtered, $post, $language, $is_translatable ) {
```
Comment (manooweb) [inline tests/phpunit/tests/test-preload-paths.php:70]: Typo

```suggestion
	public function test_preload_paths_in_widget_editor_context( $path, $is_filtered, $post, $language, $is_translatable ) {
```

**Review by manooweb — CHANGES_REQUESTED** [review]

**Review by manooweb — APPROVED** [review]

**Review by Chouby — CHANGES_REQUESTED** [review]

File: admin/admin-block-editor.php [hunk 14]
```diff
@@ -15,56 +15,75 @@ class PLL_Admin_Block_Editor {
...
+		if ( ! $context instanceof WP_Block_Editor_Context ) {
+			return $preload_paths;
+		}
+
+		// Backward compatibility with WP < 6.0 where `WP_Block_Editor_Context::$name` doesn't exist yet.
 		if (
-			$context instanceof WP_Block_Editor_Context
-			&& $context->post instanceof WP_Post
-			&& $this->model->is_translated_post_type( $context->post->post_type )
+			( property_exists( $context, 'name' ) && 'core/edit-post' !== $context->name )
```
[the remark this answers is not in the document]
Comment (Hug0-Drelon) [reply admin/admin-block-editor.php:53]: As we discussed, it is required for the backward compatibility.

File: admin/admin-block-editor.php [hunk 15]
```diff
@@ -15,56 +15,75 @@ class PLL_Admin_Block_Editor {
...
+		}
+
+		if ( ! $this->model->is_translated_post_type( $context->post->post_type ) ) {
+			return $preload_paths;
+		}
+
+		// Set default language according to the context if no language is defined yet.
+		if ( ! empty( $this->posts ) ) {
+			$this->posts->set_default_language( $context->post->ID );
+		}
```
[the remark this answers is not in the document]
Comment (Hug0-Drelon) [reply admin/admin-block-editor.php]: Indeed, removed in 00d72bea.

File: tests/phpunit/includes/testcase-preload-paths.php [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+	/**
+	 * @dataProvider preload_paths_provider
+	 *
+	 * @param string|array $path            The preload path under test. Could be an array if provided along a HTTP method.
+	 * @param bool         $is_filtered     Whether the path should be filtered or not.
+	 * @param WP_Post      $post            The post provided for the context.
+	 * @param string       $language        The post's language slug, empty if none.
+	 * @param bool         $is_translatable Whether or not the post type is translatable.
+	 */
+	abstract public function test_preload_paths_in_widget_editor_context( $path, $is_filtered, $post, $language, $is_translatable );
```
[the remark this answers is not in the document]
Comment (Hug0-Drelon) [reply tests/phpunit/includes/testcase-preload-paths.php]: Not useful, removed in 00d72bea.

**Pull request merged**